RULES/hdl/questa.go

```diff
@@ -2,75 +2,573 @@ package hdl
+				cmd = cmd + " -suppress 2583 -svinputport=net"
+				cmd = cmd + fmt.Sprintf(" +incdir+%s", core.SourcePath("").String())
+				for _, inc := range incs {
+					cmd = cmd + fmt.Sprintf(" +incdir+%s", path.Dir(inc.Absolute()))
+				}
+			} else if IsVhdl(src.String()) {
+				tool = "vcom"
+			}
+
+			// Remove the log file if the command fails to ensure we can recompile it
+			cmd = tool + " " + cmd + " " + src.String() + " || rm " + log.String()
+
+			// Add the compilation command as a build step with the log file as the
+			// generated output
+			ctx.AddBuildStep(core.BuildStep{
+				Out:   log,
+				Ins:   append(deps, src),
+				Cmd:   cmd,
+				Descr: fmt.Sprintf("%s: %s", tool, src.Relative()),
+			})
+
+			// Add the log file to the dependencies of the next files
+			deps = append(deps, log)
+
+			// Note down the created rule
+			rules[log.String()] = true
+		}
+	}
+
+	return deps, incs
+}
+
+// compileIp compiles the IP dependencies and the source files and an IP.
+func compileIp(ctx core.Context, rule Simulation, ip Ip,
+	deps []core.Path, incs []core.Path) ([]core.Path, []core.Path) {
+	for _, sub_ip := range ip.Ips() {
+		deps, incs = compileIp(ctx, rule, sub_ip, deps, incs)
+	}
+	deps, incs = compileSrcs(ctx, rule, deps, incs, ip.Sources())
+
+	return deps, incs
+}
+
+// compile compiles the IP dependencies and source files of a simulation rule.
+func compile(ctx core.Context, rule Simulation) []core.Path {
+	incs := []core.Path{}
+	deps := []core.Path{}
+
+	for _, ip := range rule.Ips {
+		deps, incs = compileIp(ctx, rule, ip, deps, incs)
+	}
+	deps, incs = compileSrcs(ctx, rule, deps, incs, rule.Srcs)
+
+	return deps
+}
+
+// optimize creates and optimized version of the design optionally including
+// coverage recording functionality. The optimized design unit can then conveniently
+// be simulated using 'vsim'.
+func optimize(ctx core.Context, rule Simulation, deps []core.Path) {
+	top := "board"
+	if rule.Top != "" {
+		top = rule.Top
+	}
+
+	cover_flag := ""
+	log_file_suffix := "vopt.log"
+	if Coverage.Value() {
+		cover_flag = "+cover"
+		log_file_suffix = "vopt_cover.log"
+	}
+
+	log_files := []core.OutPath{}
+	targets := []string{}
+	params := []string{}
+	if rule.Params != nil {
+		for key, _ := range rule.Params {
+			log_files = append(log_files, rule.Path().WithSuffix("/"+key+"_"+log_file_suffix))
+			targets = append(targets, rule.Target()+key)
+			params = append(params, key)
+		}
+	} else {
+		log_files = append(log_files, rule.Path().WithSuffix("/"+log_file_suffix))
+		targets = append(targets, rule.Target())
+		params = append(params, "")
+	}
+
+	for i := range log_files {
+		log_file := log_files[i]
+		target := targets[i]
+		param_set := params[i]
+
+		// Skip if we already have a rule
+		if rules[log_file.String()] {
+			return
+		}
+
+		cmd := fmt.Sprintf("vopt %s %s +acc=%s -l %s -work %s %s -o %s",
+			common_flags, cover_flag, Access.Value(),
+			log_file.String(), rule.Lib(), top, target)
```

Comment:
I'm getting the following error message: 
```
vopt -nologo -quiet  +acc=rna -l /home/ng/code/exp-compute/BUILD/OUTPUT-EEC93580/dta_core/vopt.log -work dta_coreLib board -o dta_core
** Error: (vopt-1927) "dta_core" cannot be used as an optimized design name because it is the name of an existing design unit.
```
You probably have to change the target name of the optimized design: With 
```
cmd := fmt.Sprintf("vopt %s %s +acc=%s -l %s -work %s %s -o %s_opt",
```
The command succeeds. You'll have to adapt the vsim invokation too.

Reply:
fixed.